File: Insert Interval/insertInterval.js
function insertInterval(intervals, newInterval) {
    let result = []
    let isFinished = false
    
    for(let i = 0; i < intervals.length; i++) {
        const interval = intervals[i]
        
        if(newInterval[0] > interval[1]) {
            result.push(interval)
        } else if(newInterval[1] < interval[0]) {
            result = result.concat([newInterval], intervals.slice(i))
            isFinished = true
            break;
        } else {
            newInterval[0] = Math.min(newInterval[0], interval[0])
            newInterval[1] = Math.max(newInterval[1], interval[1])
        }
    }
    
    if(!isFinished) result.push(newInterval)
    
    return result
}

/*
  intervals = [[1, 3], [6, 9]]
  newInterval = [2, 5]
  console.log(insertInterval(intervals, newInterval)) // [[1, 5], [6, 9]]
*/

File: Insert-Interval/Insert-Interval.java
 class Solution 
{
    public int[][] insert(int[][] intervals, int[] newInterval) 
    {
        int n = intervals.length;
        List<int[]> r = new ArrayList();

        for (int[] inter : intervals)
        {
            if (is_overlap(inter, newInterval))
            {
                newInterval[0] = Math.min(newInterval[0], inter[0]);
                newInterval[1] = Math.max(newInterval[1], inter[1]);
            }
            else
            {
                r.add(inter);
            }
        }
        
        r.add(newInterval);
        int[][] t = new int[r.size()][2];
        for (int i = 0; i < t.length; ++i) t[i] = r.get(i);
        Arrays.sort(t, (a, b) -> a[0] - b[0]);
        return t;
    }
    
    boolean is_overlap(int[] a, int[] b)
    {
        return !(a[1] < b[0] || a[0] > b[1]);
    }
}

    
    
